jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/queue/QueueStatistics.kt

```diff
@@ -0,0 +1,227 @@
+            val now = System.currentTimeMillis()
+            stats["added"] = totalPacketsAdded.sum()
+            stats["removed"] = totalPacketsRemoved.sum()
+            stats["dropped"] = totalPacketsDropped.sum()
+            val duration = (now - firstPacketAddedMs) / 1000.0
+            stats["duration_s"] = duration
+            val packetsRemoved = totalPacketsRemoved.sum().toDouble()
+            stats["average_remove_rate_pps"] = packetsRemoved / duration
+            stats["queue_size_at_remove"] = queueLengthStats.toJson()
+            queueWaitStats?.let { stats["queue_wait_time"] = it.toJson() }
+            return stats
+        }
+
+    /**
+     * Registers the addition of a packet.
+     */
+    fun added(now: Long) {
+        if (firstPacketAddedMs < 0) {
+            firstPacketAddedMs = now
+        }
+        totalPacketsAdded.increment()
+    }
+
+    /**
+     * Registers the removal of a packet.
+     */
+    fun removed(now: Long, queueSize: Int, waitTime: Long?) {
+        totalPacketsRemoved.increment()
+        queueLengthStats.addValue(queueSize.toLong())
+        if (waitTime != null) {
+            queueWaitStats?.addValue(waitTime)
+        }
+    }
+
+    /**
+     * Registers that a packet was dropped.
+     */
+    fun dropped(now: Long) {
+        totalPacketsDropped.increment()
+    }
+
+    companion object {
+        /**
+         * The scale to use for [RateStatistics]. 1000 means units
+         * (e.g. packets) per second.
+         */
+        private const val SCALE = 1000
+
+        /**
+         * The interval for which to calculate rates in milliseconds.
+         */
+        private const val INTERVAL_MS = 5000
+
+        /**
+         * Calculate the statistics buckets for a given queue length
+         */
+        private fun lengthBucketSizes(size: Int): LongArray {
+            val list = ArrayList<Long>()
+            list.add(0L)
+            var i = 1L
+
+            while (i < size) {
+                list.add(i)
+                i *= 4
+            }
+            val half = (size / 2).toLong()
+            if (half > list.last()) {
+                list.add(half)
+            }
+
+            val threeQuarters = (size * 3 / 4).toLong()
+            if (threeQuarters > list.last()) {
+                list.add(threeQuarters)
+            }
+
+            return list.toLongArray()
+        }
+
+        /**
+         * The queue waiting time bucket sizes.
+         */
+        private val waitBucketSizes = longArrayOf(2, 5, 20, 50, 200, 500, 1000)
+    }
+}
+
+class QueueStatistics(
+    /**
+     * The queue being monitored
+     */
+    private val queue: PacketQueue<*>
+) : PacketQueue.Observer {
+    /**
+     * A map of the time when objects were put in the queue
+     */
+    private val insertionTime = if (trackTimes) Collections.synchronizedMap(IdentityHashMap<Any, Long>()) else null
+
+    private val specific = if (debug) QueueStatisticsInternal(queue.capacity()) else null
```

Comment:
`specific` means "stats for this specific queue"?  How about `localStats` and `globalStats` for these vars?

Reply:
Works for me.